hw/ip/otbn/util/shared/instruction_count_range.py

```diff
@@ -0,0 +1,157 @@
+                )
+            loc_min, loc_max = 0, 0
+        elif isinstance(loc, LoopEnd):
+            # All LoopEnds should have been handled above!
+            raise ValueError(
+                f'Unexpected loop end at PC {section.end:#x} when stopping point was {stop_at}.'
+            )
+        elif isinstance(loc, LoopStart):
+            # Calculate the number of iterations if possible.
+            insn = program.get_insn(section.end)
+            if insn.mnemonic == 'loopi':
+                op_vals = program.get_operands(section.end)
+                num_iterations = op_vals['iterations']
+                loop_min, loop_max = _get_insn_count_range(
+                    program, graph, loc.loop_start_pc, 'loop end')
+                loop_min *= num_iterations
+                loop_max *= num_iterations
+            else:
+                # Cannot determine # iterations statically.
+                loop_min, loop_max = 0, None
+
+            # Calculate the instruction count range after the loop.
+            post_loop_min, post_loop_max = _get_insn_count_range(
+                program, graph, loc.loop_end_pc + 4, stop_at)
+            loc_min = loop_min + post_loop_min
+            if loop_max is None or post_loop_max is None:
+                loc_max = None
+            else:
+                loc_max = loop_max + post_loop_max
+        elif isinstance(loc, Cycle):
+            # For cycles, the minimum is 0 (if the cycle is never traversed)
+            # and the maximum is None (because we can't easily calculate
+            # statically how many times the cycle will be traversed).
+            loc_min, loc_max = 0, None
+        else:
+            insn = program.get_insn(section.end)
+            operands = program.get_operands(section.end)
+            if insn.mnemonic == 'jal' and operands['grd'] == 1:
+                # Jumping to another subroutine; count the range for the
+                # subroutine itself.
+                jump_min, jump_max = _get_insn_count_range(
+                    program, graph, loc.pc, 'ret')
+                # Calculate the instruction count range after returning from
+                # the jump.
+                post_jump_min, post_jump_max = _get_insn_count_range(
+                    program, graph, section.end + 4, stop_at)
+                loc_min = jump_min + post_jump_min
+                if jump_max is None or post_jump_max is None:
+                    loc_max = None
+                else:
+                    loc_max = jump_max + post_jump_max
+            else:
+                # If not a jump, then this is just a normal PC (i.e. a branch).
+                # Follow the branch to get the min/max range.
+                loc_min, loc_max = _get_insn_count_range(
+                    program, graph, loc.pc, stop_at)
+
+        # Merge the min/max for this location into the final result
+        if min_count is None:
+            min_count = sec_count + loc_min
```

Comment:
nit: choosing the initial values as `min_count = math.inf` and `max_count =  -math.inf` would probably simplify this part.